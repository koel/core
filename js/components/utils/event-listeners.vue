<script lang="ts">
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import Vue, { VNode } from 'vue'
import isMobile from 'ismobilejs'
import router from '@/router'
import { auth } from '@/services'
import { playlistStore, preferenceStore, userStore } from '@/stores'
import { alerts, eventBus, forceReloadWindow } from '@/utils'
import { events } from '@/config'

export default Vue.extend({
  render: (h: Function): VNode => h(),
  created (): void {
    eventBus.on({
      [events.PLAYLIST_DELETE]: (playlist: Playlist): void => {
        const destroy = async (): Promise<void> => {
          await playlistStore.delete(playlist)
          alerts.success(`Deleted playlist &quot;${playlist.name}&quot;.`)
          router.go('home')
        }

        if (!playlist.songs.length) {
          destroy()
        } else {
          alerts.confirm(`Delete the playlist &quot;${playlist.name}&quot?`, destroy)
        }
      },

      /**
       * Log the current user out and reset the application state.
       */
      [events.LOG_OUT]: async (): Promise<void> => {
        await userStore.logout()
        auth.destroy()
        forceReloadWindow()
      },

      [events.KOEL_READY]: (): void => router.init(),

      /**
       * Hide the panel away if a main view is triggered on mobile.
       */
      [events.LOAD_MAIN_CONTENT]: (): void => {
        if (isMobile.phone) {
          preferenceStore.showExtraPanel = false
        }
      }
    })
  }
})
</script>
