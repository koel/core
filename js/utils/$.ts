/**
 * A utility that aims to replace jQuery for the most basic DOM methods.
 */
export const $ = {
  is: (el: Element, selector: string): boolean => {
    return el.matches(selector)
  },

  addClass: (el: Element | null, className: string): void => {
    if (el) {
      el.classList.add(className)
    }
  },

  removeClass: (el: Element | null, className: string): void => {
    if (el) {
      el.classList.remove(className)
    }
  },

  scrollTo (el: Element, to: number, duration: number, cb?: Function): void {
    if (duration <= 0 || !el) {
      return
    }

    const difference = to - el.scrollTop
    const perTick = difference / duration * 10

    window.setTimeout((): void => {
      el.scrollTop = el.scrollTop + perTick

      if (el.scrollTop === to) {
        cb && cb()
        return
      }

      this.scrollTo(el, to, duration - 10)
    }, 10)
  }
}
