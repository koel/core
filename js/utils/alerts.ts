import alertify from 'alertify.js'

export const alerts = {
  alert: (msg: string): void => alertify.alert(msg),
  confirm: (msg: string, okFunc: Function, cancelFunc?: Function): void => {
    alertify.confirm(msg, okFunc, cancelFunc)
  },

  log: (msg: string, type: string, cb?: Function): void => {
    alertify.logPosition('top right')
    alertify.closeLogOnClick(true)

    switch (type) {
      case 'success':
        alertify.success(msg, cb)
        break
      case 'error':
        alertify.error(msg, cb)
        break
      default:
        alertify.log(msg, cb)
        break
    }
  },

  success (msg: string, cb?: Function): void {
    this.log(msg, 'success', cb)
  },

  error (msg: string, cb?: Function): void {
    this.log(msg, 'error', cb)
  }
}
