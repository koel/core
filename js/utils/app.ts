// App (electron)-only methods
let mainWindow: any

if (KOEL_ENV === 'app') {
  mainWindow = require('electron').remote.getCurrentWindow()
}

export const app = {
  triggerMaximize: (): void =>
    mainWindow && (mainWindow.isMaximized() ? mainWindow.unmaximize() : mainWindow.maximize())
}
