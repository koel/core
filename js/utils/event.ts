import event from '@phanan/vuebus'
import { events as eventNames } from '@/config'

(<any>event).$names = eventNames

export { event }
