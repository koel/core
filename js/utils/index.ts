export * from './event'
export * from './alerts'
export * from './filters'
export * from './formatters'
export * from './supports'
export * from './common'
export * from './$'
export * from './app'
export * from './helpers'
export * from './file-reader'
