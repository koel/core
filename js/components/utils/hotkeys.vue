<template>
  <global-events
    @keydown.space = "togglePlayback"
    @keydown.j = "playNext"
    @keydown.k = "playPrev"
    @keydown.f = "search"
    @keydown.l = "toggleLike"
    @keydown.mediaPrev = "playPrev"
    @keydown.mediaNext = "playNext"
    @keydown.mediaToggle = "togglePlayback"
  />
</template>

<script lang="ts">
import Vue from 'vue'
import GlobalEvents from 'vue-global-events'
import { $, event, noop } from '@/utils'
import { playback, socket } from '@/services'
import { queueStore, favoriteStore, songStore } from '@/stores'

let ipc: any, events: any

if (KOEL_ENV === 'app') {
  ipc = require('electron').ipcRenderer
  events = require('&/events').default
}

// Register our custom key codes
Vue.config.keyCodes = {
  a: 65,
  j: 74,
  k: 75,
  f: 70,
  l: 76,
  mediaNext: 176,
  mediaPrev: 177,
  mediaToggle: 179
}

/**
 * Listen to the global shortcuts (media keys).
 * Only works in the app.
 */
let listenToGlobalShortcuts = noop

if (KOEL_ENV === 'app') {
  listenToGlobalShortcuts = (): void => {
    const mediaFunctionMap = {
      MediaNextTrack: () => playback.playNext(),
      MediaPreviousTrack: () => playback.playPrev(),
      MediaStop: () => playback.stop(),
      MediaPlayPause: () => playback.toggle()
    } as { [propName: string]: Function }

    ipc.on(events.GLOBAL_SHORTCUT, (e: any, msg: string) => msg in mediaFunctionMap && mediaFunctionMap[msg]())
  }
}

export default Vue.extend({
  components: {
    GlobalEvents
  },

  methods: {
    /**
     * Toggle playback when user presses Space key.
     *
     * @param {Object} e The keydown event
     */
    togglePlayback: (e: KeyboardEvent): boolean => {
      if (
        !(e.currentTarget instanceof Document) &&
        $.is(e.currentTarget as Element, 'input, textarea, button, select') &&
        !$.is(e.currentTarget as Element, 'input[type=range]')
      ) {
        return true
      }

      e.preventDefault()
      playback.toggle()

      return false
    },

    /**
     * Play the previous song when user presses K.
     */
    playPrev: (e: KeyboardEvent): boolean => {
      if (
        !(e.currentTarget instanceof Document) &&
        $.is(e.currentTarget as Element, 'input, select, textarea')
      ) {
        return true
      }

      e.preventDefault()
      playback.playPrev()

      return false
    },

    /**
     * Play the next song when user presses J.
     */
    playNext: (e: KeyboardEvent): boolean => {
      if (
        !(e.currentTarget instanceof Document) &&
        $.is(e.currentTarget as Element, 'input, select, textarea')
      ) {
        return true
      }

      e.preventDefault()
      playback.playNext()

      return false
    },

    /**
     * Put focus into the search field when user presses F.
     */
    search: (e: KeyboardEvent): boolean => {
      if (
        !(e.currentTarget instanceof Document) &&
        ($.is(e.currentTarget as Element, 'input, select, textarea') &&
        !$.is(e.currentTarget as Element, 'input[type=range]'))
      ) {
        return true
      }

      if (e.metaKey || e.ctrlKey) {
        return true
      }

      e.preventDefault()
      event.emit(event.$names.FOCUS_SEARCH_FIELD)

      return false
    },

    /**
     * Like/unlike the current song when use presses L.
     */
    toggleLike: (e: KeyboardEvent): boolean => {
      if (
        !(e.currentTarget instanceof Document) &&
        $.is(e.currentTarget as Element, 'input, select, textarea')
      ) {
        return true
      }

      if (!queueStore.current) {
        return false
      }

      favoriteStore.toggleOne(queueStore.current)
      socket.broadcast(event.$names.SOCKET_SONG, songStore.generateDataToBroadcast(queueStore.current))
      return false
    }
  },

  created: (): void => {
    if (KOEL_ENV === 'app') {
      listenToGlobalShortcuts()
    }
  }
})
</script>
