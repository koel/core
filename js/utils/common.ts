import select from 'select'
import { event, noop, pluralize } from '@/utils'
import { sharedStore } from '@/stores'

/**
 * Load (display) a main panel (view).
 *
 * @param {...*} args     Extra data to attach to the view.
 */
export const loadMainView = (view: MainViewName, ...args: any[]): void =>
  event.emit(event.$names.LOAD_MAIN_CONTENT, view, ...args)

/**
 * Force reloading window regardless of "Confirm before reload" setting.
 * This is handy for certain cases, for example Last.fm connect/disconnect.
 */
export const forceReloadWindow = (): void => {
  if (window.__UNIT_TESTING__) {
    return
  }

  window.onbeforeunload = noop
  window.location.reload()
}

export const showOverlay = (message = 'Just a little patienceâ€¦', type = 'loading', dismissable = false): void =>
  event.emit(event.$names.SHOW_OVERLAY, { message, type, dismissable })

export const hideOverlay = (): void => event.emit(event.$names.HIDE_OVERLAY)

export const copyText = (text: string): void => {
  let copyArea = document.querySelector('#copyArea') as HTMLTextAreaElement

  if (!copyArea) {
    copyArea = document.createElement('textarea')
    copyArea.id = 'copyArea'
    document.body.appendChild(copyArea)
  }

  copyArea.style.top = `${window.pageYOffset || document.documentElement.scrollTop}px`
  copyArea.value = text
  select(copyArea)
  document.execCommand('copy')
}

export const getDefaultCover = (): string => `${sharedStore.state.cdnUrl}/public/img/covers/unknown-album.png`

const createGhostDragImage = (event: DragEvent, text: string): void => {
  if (!event.dataTransfer) {
    return
  }

  let dragGhost = document.querySelector('#dragGhost') as HTMLElement

  if (!dragGhost) {
    // Create the element to be the ghost drag image.
    dragGhost = document.createElement('div')
    dragGhost.id = 'dragGhost'
    document.body.appendChild(dragGhost)
  }

  dragGhost.innerText = text
  event.dataTransfer.setDragImage(dragGhost, 0, 0)
}

/**
 * Handle song/album/artist drag start event.
 * @param {Song|Song[]} dragged Either an array of songs or a song/album/artist object
 */
export const startDragging = (event: DragEvent, dragged: Song | Song[] | Album | Artist, type: DragType): void => {
  if (!event.dataTransfer) {
    return
  }

  let text
  let songIds

  switch (type) {
    case 'Song':
      dragged = (<Song[]>[]).concat(<Song>dragged)
      text = dragged.length === 1
        ? `${dragged[0].title} by ${dragged[0].artist.name}`
        : pluralize(dragged.length, 'song')
      songIds = dragged.map(song => song.id)
      break

    case 'Album':
      dragged = <Album>dragged
      text = `All ${pluralize(dragged.songs.length, 'song')} in ${dragged.name}`
      songIds = dragged.songs.map(song => song.id)
      break

    case 'Artist':
      dragged = <Artist>dragged
      text = `All ${pluralize(dragged.songs.length, 'song')} by ${dragged.name}`
      songIds = dragged.songs.map(song => song.id)
      break

    default:
      throw Error(`Invalid drag type: ${type}`)
  }

  event.dataTransfer.setData('application/x-koel.text+plain', songIds.join(','))
  event.dataTransfer.effectAllowed = 'move'

  createGhostDragImage(event, text)
}
