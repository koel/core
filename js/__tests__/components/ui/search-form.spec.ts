import Component from '@/components/ui/search-form.vue'
import { mock } from '@/__tests__/__helpers__'
import { shallow } from '@/__tests__/adapter'
import { eventBus } from '@/utils'

describe('components/ui/search-form', () => {
  afterEach(() => {
    jest.resetModules()
    jest.clearAllMocks()
  })

  it('renders properly', () => {
    expect(shallow(Component)).toMatchSnapshot()
  })

  it('emits an event to filter', async done => {
    const emitStub = mock(eventBus, 'emit')
    const wrapper = shallow(Component)
    wrapper.find('[type=search]').setValue('foo').input()

    setTimeout(() => {
      expect(emitStub).toHaveBeenCalledWith('SEARCH_KEYWORDS_CHANGED', 'foo')
      done()
    }, 500) // because of debounce
  })
})
