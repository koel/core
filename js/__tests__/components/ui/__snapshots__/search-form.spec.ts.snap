// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ui/search-form renders properly 1`] = `<div id="searchForm" role="search" class="side search showing"><input type="search" autocorrect="false" placeholder="Press F to search" spellcheck="false" class=""></div>`;
